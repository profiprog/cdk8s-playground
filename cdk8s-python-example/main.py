#!/usr/bin/env python
from constructs import Construct
from cdk8s import App, Chart
from imports import k8s

class MyChart(Chart):
    def __init__(self, scope: Construct, ns: str):
        super().__init__(scope, ns)

        labels = {"app": "cdk8s-python"}

        k8s.Service(
            self,
            'service',
            spec=k8s.ServiceSpec(
                type='LoadBalancer',
                ports=[
                    k8s.ServicePort(
                        port=80,
                        target_port=k8s.IntOrString.from_number(80)
                    ),
                ],
                selector=labels,
            ),
        )

        k8s.Deployment(
            self,
            'deployment',
            spec=k8s.DeploymentSpec(
                replicas=2,
                selector=k8s.LabelSelector(match_labels=labels),
                template=k8s.PodTemplateSpec(
                    metadata=k8s.ObjectMeta(labels=labels),
                    spec=k8s.PodSpec(
                        containers=[
                            k8s.Container(
                                name=ns,
                                image='nginxdemos/hello',
                                ports=[
                                    k8s.ContainerPort(container_port=80),
                                ],
                            ),
                        ],
                    ),
                ),
            ),
        )


app = App()
MyChart(app, "cdk8s-python-example")

app.synth()
