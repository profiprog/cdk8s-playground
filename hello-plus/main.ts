import { Construct } from 'constructs';
import { ApiObjectMetadata, App, Chart } from 'cdk8s';
import { Container, Deployment, ImagePullPolicy, Ingress, IngressBackend, Protocol, Service, ServiceType } from 'cdk8s-plus';
import { Namespace } from './imports/k8s';

export interface SimpleWebAppChartOptions {
  domain?: string;
  appImage: string;
  /** default 80 */
  cotnainerPort?: number;
  servicePort: number;
  /** default 1 */
  replicas?: number;
  namespace?: string;
}

export class SimpleWebAppChart extends Chart {
  constructor(scope: Construct, name: string, opts: SimpleWebAppChartOptions) {
    super(scope, name);

    let metadata: ApiObjectMetadata = {};
    if (opts.namespace) metadata = {...metadata, namespace: opts.namespace};

    const service = new Service(this, 'service', {
      metadata,
      type: ServiceType.LOAD_BALANCER,
      ports: [{port: opts.servicePort, targetPort: opts.cotnainerPort || 80, protocol: Protocol.TCP }],
    });
    service.addSelector('app', `web-${name}`);

    if (opts.domain) {
      new Ingress(this, 'ingress', {
        metadata,
        rules: [
          { host: opts.domain, path: '/', backend: IngressBackend.fromService(service) },
        ],
      });
    }

    const deployment = new Deployment(this, 'deployment', {
      metadata,
      replicas: opts.replicas || 1,
      defaultSelector: false,
      containers: [
        new Container({
          image: opts.appImage,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          name: name,
          port: opts.cotnainerPort || 80,
        }),
      ],
    });
    deployment.podMetadata.addLabel('app', `web-${name}`)
    deployment.selectByLabel('app', `web-${name}`);

  }
}

export class NamespaceChart extends Chart {
  private readonly _namespace: Namespace;

  constructor(scope: Construct, name: string) {
    super(scope, name);

    this._namespace = new Namespace(this, 'namespace', { metadata: { name } });
  }

  getNamespace(): Namespace {
    return this._namespace;
  }

  get name(): string {
    return this._namespace.name;
  }
}


const app = new App();
const ns = new NamespaceChart(app, 'web-apps');

new SimpleWebAppChart(app, 'hellok8s', {
  appImage: 'paulbouwer/hello-kubernetes:1.7',
  cotnainerPort: 8080,
  servicePort: 9001,
  replicas: 2,
  domain: 'k8s.local',
  namespace: ns.name,
}).addDependency(ns);

new SimpleWebAppChart(app, 'nginxdemo', {
  appImage: 'nginxdemos/hello',
  servicePort: 9002,
  replicas: 2,
  domain: 'nginx.local',
  namespace: ns.name,
}).addDependency(ns);

new SimpleWebAppChart(app, 'realcount', {
  appImage: 'tutum/hello-world',
  servicePort: 9003,
  replicas: 2,
  domain: 'tutum.local',
  namespace: ns.name,
}).addDependency(ns);

app.synth();
