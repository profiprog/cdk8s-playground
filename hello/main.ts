import { Construct } from 'constructs';
import { App, Chart } from 'cdk8s';

import { Deployment, Service, Ingress, IntOrString, Namespace, ObjectMeta } from './imports/k8s'

export interface SimpleWebAppOption {
  domain?: string;
  appImage: string;
  /** default 80 */
  cotnainerPort?: number;
  servicePort: number;
  /** default 1 */
  replicas?: number;
  namespace?: string;
}

export class SimpleWebApp extends Chart {
  constructor(scope: Construct, name: string, opts: SimpleWebAppOption) {
    super(scope, name);

    const label = { app: `web-${name}` };

    let metadata: ObjectMeta = {};
    if (opts.namespace) metadata = {...metadata, namespace: opts.namespace};

    const service = new Service(this, 'service', {
      metadata,
      spec: {
        type: 'LoadBalancer',
        ports: [
          {
            port: opts.servicePort,
            targetPort: IntOrString.fromNumber(opts.cotnainerPort || 80),
            protocol: 'TCP',
          },
        ],
        selector: label,
      },
    });

    if (opts.domain) {
      new Ingress(this, 'ingress', {
        metadata,
        spec: {
          rules: [
            {
              host: opts.domain,
              http: {
                paths: [
                  {
                    path: '/',
                    backend: {
                      serviceName: service.name,
                      servicePort: opts.servicePort,
                    },
                  },
                ],
              },
            },
          ],
        },
      });
    }

    new Deployment(this, 'deployment', {
      metadata,
      spec: {
        replicas: opts.replicas || 1,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: label },
          spec: {
            containers: [
              {
                name: name,
                image: opts.appImage,
                ports: [ { containerPort: opts.cotnainerPort || 80 } ],
              },
            ],
          },
        },
      }
    });
  }
}



export class NamespaceChart extends Chart {
  private readonly _namespace: Namespace;

  constructor(scope: Construct, name: string) {
    super(scope, name);

    this._namespace = new Namespace(this, 'namespace', { metadata: { name } });
  }

  getNamespace(): Namespace {
    return this._namespace;
  }

  get name(): string {
    return this._namespace.name;
  }
}




const app = new App();
const ns = new NamespaceChart(app, 'web-apps');

new SimpleWebApp(app, 'hellok8s', {
  appImage: 'paulbouwer/hello-kubernetes:1.7',
  cotnainerPort: 8080,
  servicePort: 9001,
  replicas: 2,
  domain: 'k8s.local',
  namespace: ns.name,
}).addDependency(ns);

new SimpleWebApp(app, 'nginxdemo', {
  appImage: 'nginxdemos/hello',
  servicePort: 9002,
  replicas: 2,
  domain: 'nginx.local',
  namespace: ns.name,
}).addDependency(ns);

new SimpleWebApp(app, 'realcount', {
  appImage: 'tutum/hello-world',
  servicePort: 9003,
  replicas: 2,
  domain: 'tutum.local',
  namespace: ns.name,
}).addDependency(ns);

app.synth();
