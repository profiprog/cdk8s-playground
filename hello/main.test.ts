import {SimpleWebApp} from './main';
import {Testing} from 'cdk8s';

describe('Placeholder', () => {
  test('Empty', () => {
    const app = Testing.app();
    const chart = new SimpleWebApp(app, 'test-app', {
      appImage: 'test-image',
      servicePort: 80,
      domain: 'test.local',
    });
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
