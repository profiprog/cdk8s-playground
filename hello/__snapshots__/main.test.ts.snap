// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder Empty 1`] = `
Array [
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "name": "test-chart-app-025556c3",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "test.local",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "service-fjdasf",
                  "servicePort": 8099,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "service-fjdasf",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 8099,
          "protocol": "TCP",
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "app": "hello-ckd8s",
      },
      "type": "LoadBalancer",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-chart-deployment-a0c86b87",
    },
    "spec": Object {
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "app": "hello-ckd8s",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "hello-ckd8s",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "nginxdemos/hello",
              "name": "hello-kubernetes",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
]
`;
